### DISCUSSION TITLE
Flows: `{{ }}` vs `"{{ }}"` .. What's the difference?

### DISCUSSION THREAD

**Discussion by erondpowell** [directus/directus#15945] [Archive]
I noticed that you can use double mustache syntax in two ways.

`{{ $last }}` yields a result like this:

```
{
  "message": {
    "id": 109,
    "title": "A title",
    "body": "A body"
  }
}
```

but if you wrap it in quotes `"{{ $last }}"`  then it appears a bunch more quotes and line-breaks are added and the pretty JSON nesting is removed.
```
{
  "message": "\"{\"id\":109,\"title\":\"A title\",\"body\":\"A body\"}\""
}
```

This seems like some kind of stringification technique *(which makes sense bc I wrapped it in strings)* but I haven't seen it before.
What's going on with those `\`? They seem to represent line-breaks... but also end of json keys/values.. but also other stuff.

@licitdev @br41nslug

**Comment by azrikahar** [comment]
I believe it's not exactly intended based on the conversation here https://github.com/directus/directus/discussions/15025#discussioncomment-3521216, but what's happening here is they are indeed being stringified as you mentioned. The `\` in this case is used to escape the quotes. If it's `"""` for example, the 2nd quote will be interpreted as the ending quote, so the 3rd quote becomes "orphaned" and can cause error. Whereas if it's `"\""`, then the 2nd quote is escaped and won't be interpreted as the ending quote. Hence why in your example, all of the inner quotes are escaped with a backslash so none of them gets interpreted as the closing quote. `\n` would more typically be the escape sequence for line-breaks. The ["Single character escape sequences" section here](https://mathiasbynens.be/notes/javascript-escapes) may also be helpful.

Coincidentally this PR #15946 by @nickrum an hour after this discussion was posted is meant to remove this exact confusion so they will no longer stringified when wrapped in quotes (if I understood it correctly).

**Reply by erondpowell** [reply]
Thx azzy!

**Reply by nickrum** [reply]
@azrikahar is completely right 👍

To add to what he said, the initial idea was to be able to use

```
{
  "message": {{ $last }}
}
```

to be able to pass data that is an object or an array to an Operation. The Flows system replaces `{{ $last }}` with the JSON representation of what's in `$last`, which results in

```json
{
  "message": { "id": 109, "title": "A title", "body": "A body" }
}
```

, parses the JSON to a Javascript object and passes it to the Operation. The problem with this is that `{ "message": {{ $last }} }` is not valid JSON, which means you need to use the raw text editor to input this. That's why most users tried to do

```json
{
  "message": "{{ $last }}"
}
```

The Flows system will replace `{{ $last }}` with the JSON representation of what's in `$last`, which then results in

```json
{
  "message": "\"{\"id\":109,\"title\":\"A title\",\"body\":\"A body\"}\""
}
```

Parsing this to a Javascript object will result in an object with a single `message` property that contains a string with the JSON representation of `$last`, which is likely not what the user intended to get.

In addition to the option to pass options as JSON, which are then converted back to Javascript objects, the Flows system also accepts raw Javascript objects. This is (confusingly) what is being done when using the JSON code editor to input options. The App will parse the JSON to a Javascript object before passing it to the API and the Flows system. That's why it is not possible to pass the invalid JSON string `{ "message": {{ $last }} }` to the JSON code editor and the user having to use the raw text editor. The option will allways look like this:

```js
{
  message: "\"{\"id\":109,\"title\":\"A title\",\"body\":\"A body\"}\""
}
```

Normally, the Flows system converts any value that it encounters inside `{{  }}` in a template to a string by converting it to JSON because it has to somehow include non-string values inside a sting template.
Before #15946, there was a single exception to this: When passing a template with a single value inside (e.g. `"{{ $last }}"`) to an option, the Flows system will not stringify `$last` to JSON only to parse it back to an Javascript object in the next step. Instead it will pass `$last` as is to the Operation.
With #15946, the same thing is now done for all values inside objects or arrays, which means the Javascript object

```js
{
  message: "{{ $last }}"
}
```

is now converted to

```js
{
  message: { id: 109, title: 'A title', body: 'A body' }
}
```

and

```js
{
  message: "Message: {{ $last }}"
}
```

is still converted to

```js
{
  message: "Message: {\"id\":109,\"title\":\"A title\",\"body\":\"A body\"}"
}
```

This should make it a lot easier to pass object or array data between Operations without any real downsides (I hope). Also, sorry for the very vague description in #15946.
